numpy/core/src/npysort/selection.c.src

```diff
@@ -1,419 +0,0 @@
-/* -*- c -*- */
-
-/*
- *
- * The code is loosely based on the quickselect from
- * Nicolas Devillard - 1998 public domain
- * http://ndevilla.free.fr/median/median/
- *
- * Quick select with median of 3 pivot is usually the fastest,
- * but the worst case scenario can be quadratic complexity,
- * e.g. np.roll(np.arange(x), x / 2)
- * To avoid this if it recurses too much it falls back to the
- * worst case linear median of median of group 5 pivot strategy.
- */
-
-
-#define NPY_NO_DEPRECATED_API NPY_API_VERSION
-
-#include "npy_sort.h"
-#include "npysort_common.h"
-#include "numpy/npy_math.h"
-#include "npy_partition.h"
-#include <stdlib.h>
-
-#define NOT_USED NPY_UNUSED(unused)
-
-
-/*
- *****************************************************************************
- **                            NUMERIC SORTS                                **
- *****************************************************************************
- */
-
-
-static NPY_INLINE void store_pivot(npy_intp pivot, npy_intp kth,
-                                   npy_intp * pivots, npy_intp * npiv)
-{
-    if (pivots == NULL) {
-        return;
-    }
-
-    /*
-     * If pivot is the requested kth store it, overwriting other pivots if
-     * required. This must be done so iterative partition can work without
-     * manually shifting lower data offset by kth each time
-     */
-    if (pivot == kth && *npiv == NPY_MAX_PIVOT_STACK) {
-        pivots[*npiv - 1] = pivot;
-    }
-    /*
-     * we only need pivots larger than current kth, larger pivots are not
-     * useful as partitions on smaller kth would reorder the stored pivots
-     */
-    else if (pivot >= kth && *npiv < NPY_MAX_PIVOT_STACK) {
-        pivots[*npiv] = pivot;
-        (*npiv) += 1;
-    }
-}
-
-/**begin repeat
- *
- * #TYPE = BOOL, BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
- *         LONGLONG, ULONGLONG, HALF, FLOAT, DOUBLE, LONGDOUBLE,
- *         CFLOAT, CDOUBLE, CLONGDOUBLE#
- * #suff = bool, byte, ubyte, short, ushort, int, uint, long, ulong,
- *         longlong, ulonglong, half, float, double, longdouble,
- *         cfloat, cdouble, clongdouble#
- * #type = npy_bool, npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int,
- *         npy_uint, npy_long, npy_ulong, npy_longlong, npy_ulonglong,
- *         npy_ushort, npy_float, npy_double, npy_longdouble, npy_cfloat,
- *         npy_cdouble, npy_clongdouble#
- * #inexact = 0*11, 1*7#
```

Comment:
this inexact looks odd to me: if I understand the syntax correctly, it means the 11 first entries are exact (0) but then `npy_ushort` is flagged as inexact (1) for no apparent reason? I updated that during the conversion.

Reply:
The 16 bit floats are stored in 16 bit unsigned integers, the compiler doesn't (didn't?) have a half type.

Reply:
As a result, arithmetic for half floats is done in  float32, with conversion before and after. Comparisons may also be tricky.